caml_result: an exception monad in C for resource-safe interfaces

gasche opened the pull request:
In the OCaml runtime, some C functions that in turn call OCaml code come in two variant:
- the "raising' variant, for example `caml_callback`, will raise exceptions directly, as OCaml code would
- the 'encoded exception' variant, for example `caml_callback_exn`, will not raise exceptions but encode them into the return value

The 'encoded exception' variant is important for resource-safe interfaces because the caller may want to run some cleanup/release code before propagating the exception, which the 'raising' variant cannot express -- C code cannot create OCaml exception handlers. On the other hand, the encoding of exceptions into the `value` type is a big hack that is unsafe in two different ways:
- dynamically, the `value` encoding exceptions are not valid OCaml values, the GC will blow up if it encounters them; callers have to follow a strict discipline of checking for an exception right away and discarding the fake `value`
- statically, reusing the `value` type means that the C compiler cannot check that we use these exceptions as intended, and will happily let us mix them with real values by mistake, making the dynamic unsafety harder to guard against

In #12407, @gadmm proposed to extend the approach of 'encoded exceptions' to more parts of the runtime, to make it possible to write resource-safe code in more places, but @xavierleroy pushed against the design due to the unsafety of encoded exceptions.
To be able to move forward, we need an approach for resource-safe APIs that is consensual; the present PR, which grew out of the discussios in #12407 and in particular discussions with @gadmm, is a proposed approach.

The PR proposes a new 'result' variant: `caml_callback_result` is similar to `caml_callback_exn`, but instead of returning an unsafe `value` it returns a new, distinct C type `caml_result`, that properly represents either a value or an exception. This is more safe:
- dynamically, no invalid values are produced
- statically, the type is incompatible with `value`, reminding us to deal with the exception in the caller

```c
typedef struct {
  _Bool is_exception;
  value data;
} caml_result;
```

The present PR:
1. introoduces this `caml_result` type,
2. exposes `_result` variants for all functions that previously had an `_exn` variant, and
3. removes almost all uses of encoded exceptions in the runtime code itself.

For the `_exn` variants, the ones that were part of the public API are kept for compatibility, and those that were under CAML_INTERNALS are removed.

(The places where encoded exceptions are retained are the implementation of the `caml_callback*_exn` functions, which touch the bytecode interpreter and the assembly code in the runtime system. I preferred to leave them as-is, and wrap them inside `caml_result` on the outside.)

gasche committed:
runtime: introduce a caml_result type to supersede encoded exception values

gasche committed:
runtime: expose caml_callback*_result variants

gasche committed:
a CAMLlocalresult macro to root caml_result values

gasche committed:
remove a superfluous use of caml_callback2_exn

gasche committed:
use caml_result in domain.c

gasche committed:
use caml_callback_result when the exception is ignored

gasche committed:
convert all all remaining users of _exn variants in the runtime

(statmemprof still uses _exn internally, to be converted next)

dustanddreams commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,62 @@ typedef opcode_t * code_t;
+
+/* The [caml_result] type represents the result of computing an OCaml
+   term -- either a value or an exception:
+   - if [is_exception] is true,
+     the computation resulted in a value stored in [data]
+   - if [is_exception] is false,
+     the computation resulted in an exception stored in [data]
```
```suggestion
   - if [is_exception] is false,
     the computation resulted in a value stored in [data]
   - if [is_exception] is true,
     the computation resulted in an exception stored in [data]
```

gasche replied:
... thanks!

gasche committed:
Changes

MisterDA commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
+   lets the caller implement proper cleanup and propagate the
+   exception themselves. */
+typedef struct {
+  _Bool is_exception;
```
Note that C99 `_Bool` cannot be used in public headers for C++ compatibility as C++ doesn't define it. We chose not to include `stdbool.h` in public headers out of fear of compatibility problems, so we don't have access to `bool`. The current consensus is to use `int` for public booleans. See #12787, #12702 for previous discussions.

gadmm replied:
Note that it is also possible to test for C++ and use `bool` in this case. (This was not possible with #12787.) In the long run C is headed towards deprecating `_Bool` in favour of `bool`.

gasche replied:
It would be tempting to be able to use "`_Bool` or `bool`", because this plays nicely (nicer than `int`) with FFI based on code generation from the C headers. @gadmm tried Rust's `bindgen`, it creates a nice-looking Rust structure with a `bool` field.

However, in https://github.com/ocaml/ocaml/issues/12772#issuecomment-1824379144 @MisterDA hinted at ABI issues coming from the fact that `_Bool` and `bool` are not guaranteed to have the same size. This could be a deal-breaker, right?

I am unfamiliar with these aspects of language interoperability -- it is interesting to have to deal with the absence of a type of booleans in 2024 -- so I welcome guidance. My intuition would be to give up on the nice type and use `int` (or maybe `intnat`?) to avoid any potential issues.

NickBarnes replied:
If we're going to use a C89 type it should probably be `int`, which is (still) the type of boolean expressions (such as `(0 == 1)`, `(a && b)`, or `!p`) in C23.

gadmm replied:
Use bool/_Bool or int; it should not make any change for Rust (nor to the caml_result type due to padding). I don't think ABI issues should be feared, as OCaml is unlikely to be supported on exotic platforms/compilers.

gasche commented:
@gadmm told me (private communication) that he is fine with the overall approach if we can gather consensus around this.

damiendoligez commented on runtime/caml/callback.h:
```diff
@@ -35,9 +35,22 @@ CAMLextern value caml_callback3 (value closure, value arg1, value arg2,
+   above, but they return an 'encoded exceptional value' (see mlvalues.h)
+   which represents either a value or an exception. This interface is unsafe
+   and it is easy to make mistakes due to the lack of type information,
+   we recommend the *_result variants instead. */
```
```suggestion
   we strongly recommend the *_result variants instead. */
```

damiendoligez commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
+   bit set. These encoded exceptions are invalid values and must not
+   be seen by the garbage collector. This is unsafe, and the
+   C type-checker does not help. We strongly recommend using the
+   caml_result type above instead. Which is GC-safe and more
```
```suggestion
   caml_result type above instead, it is GC-safe and more
```

damiendoligez commented on runtime/signals.c:
```diff
@@ -350,56 +349,56 @@ value caml_do_pending_actions_exn(void)
   caml_process_external_interrupt();
 
-  return Val_unit;
+  return result;
```
```suggestion
  return Result_unit;
```
Clearer and more robust: be faithful to the original rather than rely on the fact that `caml_final_do_calls_result` happens to return `Result_unit`.

damiendoligez commented on runtime/signals.c:
```diff
@@ -350,56 +349,56 @@ value caml_do_pending_actions_exn(void)
 
-CAMLexport value caml_process_pending_actions_exn(void)
+CAMLexport caml_result caml_process_pending_actions_result(void)
 {
-  return caml_process_pending_actions_with_root_exn(Val_unit);
+  return caml_process_pending_actions_with_root_result(Val_unit);
 }
```
Too much wrapping going on here... I think this function is the same as `caml_do_pending_actions_result`.

gasche replied:
This is not just `caml_do_pending_actions_result`, as there is a check on `caml_check_pending_actions` first. I inlined/streamlined the code -- and took the opportunity to write a small bit of documentation on the `with_root` variants. This new commit will need a review because it is less of a mechanical refactoring than the rest of the patch.

damiendoligez reviewed (approved):
I like the idea and the code looks good modulo a few remarks, but don't merge before @xavierleroy has a look at it.

stedolan commented:
This looks great, the encoded exception values returned by callback_exn are an easy trap to fall into.

> (The places where encoded exceptions are retained are the implementation of the caml_callback*_exn functions, which touch the bytecode interpreter and the assembly code in the runtime system. I preferred to leave them as-is, and wrap them inside caml_result on the outside.)

This is the correct choice. The C ABI rules for returning a structure by value are tricky and vary significantly between platforms. It suits the C API (for which we have a compiler to implement those rules), but it would be unnecessarily complicated to follow these rules correctly in the per-platform assembly stubs.

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
+   - if [is_exception] is true,
+     the computation resulted in an exception stored in [data]
+
+   This is the C counterpart of a [('a, exn) result] type in
```
"counterpart" makes me think of something that it has a similar representation

gadmm commented on runtime/callback.c:
```diff
@@ -68,6 +68,14 @@ Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
+    caml_raise(res.data);
+  else
+    return res.data;
+}
```
I propose to define it right-away as Caml_inline inside e.g. `fail.h`.

gasche replied:
I wanted to expose the prototype in mlvalues, as part of the `caml_result` API, but I could define it inline there as it depends on `caml_raise`. I may revisit this once we have converged on a new name for `caml_run_result`.

gadmm commented on runtime/callback.c:
```diff
@@ -68,6 +68,14 @@ Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
 }
 
+/* support code for caml_result */
+CAMLexport value caml_run_result(caml_result res)
```
I understand the allusion to runnable monads, but a name in the same vein as `caml_raise_if_exception` would be much less cryptic.

gasche replied:
I considered the following naming choices:

1. `caml_raise_if_exception` is only in CAML_INTERNALS, so I could reuse that name and use `caml_raise_if_encoded_exception` for the old version, which is now only used in `callback.c`.
2. I could use consistent names for both versions, for example `caml_raise_if_exception_{encoded,result}`
3. or a new name altogether, for example `caml_get_value_or_raise`.

My current preference is (3) -- I like that the name is explicit that it can raise (this is better than `caml_run_result`), but also that it returns the value otherwise (this is better than `caml_raise_if_exception`).

I will move `caml_raise_if_encoded_exception` out of `fail.h` and into `callback.c` right away, and I think I will go for (3) as well, but I am interested in feedback on the name in any case.

gadmm commented on runtime/callback.c:
```diff
@@ -294,6 +302,33 @@ CAMLexport value caml_callbackN_exn(value closure, int narg, value args[]) {
 
+/* Result-returning variants of the above */
+
+CAMLexport caml_result caml_callbackN_result(
```
I am hoping to have `_result` variants for all public functions of the runtime at some point (hence for many private ones as well). Maybe some less noisy convention such as `_res` would turn out to be preferable in the long run?

gasche replied:
I prefer to have a slightly longer but clearer name. `_res` could mean various things to me without context, for example `resource` or `resolution`.

gadmm replied:
Unless it is a codebase-wide convention, and it is already clear from the signature that it returns a caml_result? In addition, to me a complete word is more at risk of introducing ambiguity with words that precede: cf. `run_callback_result` in `memprof.c`.

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
+  value data;
+} caml_result;
+
+Caml_inline caml_result Result_value(value v)
```
I am not fond of `static inline` functions that disguise as macros with their naming. I wonder what others think.

xavierleroy replied:
Independently, I wanted to suggest the use of C99 compound literals and named initializers:
```
  return (caml_result) { .is_exception = 0, .data = v };
```
Then, a macro would be just as nice.

gasche replied:
I am now using

```c
#define Result_value(v) (caml_result){ .is_exception = 0, .data = v }
#define Result_exception(exn) (caml_result){ .is_exception = 1, .data = exn }
```

Thanks for the feedback and suggestion.

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
 #define Is_exception_result(v) (((v) & 3) == 2)
 #define Extract_exception(v) ((v) & ~3)
 
+Caml_inline value Encoded_result(caml_result res) {
```
Unless you really want it public, I suggest to define it in a more specialized header or to guard it with `CAML_INTERNALS`. (Idem for `Result_encoded`.)

xavierleroy replied:
Or even don't define `Encoded_result` and `Result_encoded` in any header file and just define `Result_encoded` in the only place where it's used (runtime/callback.c).

gasche replied:
I intended to make these conversion functions public. My reasoning was the following:
- encoded exception values are part of the public API, and `caml_result` is now also part of the public API, so it makes little sense to make conversion functions internals-only
- I expect that these conversion functions will help the transition from code using encoded exceptions to code using results (and not the other way around, obviously). For example, `Encoded_result` could be used to gradually migrate producer code to use results, without breaking a public function returning an encoded exception.

(I am also reasonably confident about the interface and naming of these functions (or macros) that follow the standard convention, so making them public is not a compatibility risk.)

The sense I get from your comments is that you think no one should actually be using encoded exceptions, so we should behave as if they were private for all intent and purposes, and so there is no need to increase our API surface. Is this the idea?

xavierleroy replied:
This is exactly the idea.  By adding these `Encoded_result` and `Result_encoded` functions, you're drawing attention to the old "let's set bit number 1" API, while we want this API to sink into oblivion.  I don't think the old API was widely used by 3rd-party libraries anyway, except to interface with the `caml_callback_exn*` functions.

gadmm commented on runtime/caml/signals.h:
```diff
@@ -37,10 +37,14 @@ CAMLextern void caml_process_pending_actions (void);
+/* Returns [Val_unit] or an encoded exception.
+   Superseded by the safer [_result] variant above,
+   kept around for compatibility. */
 CAMLextern value caml_process_pending_actions_exn (void);
```
You can move the superseded functions towards the end of the public interface, and have the documentation simply say `/* Superseded by …_result */`

gadmm commented on runtime/compare.c:
```diff
@@ -99,18 +99,17 @@ static intnat compare_val(value v1, value v2, int total)
+  if (result.is_exception) {
     compare_free_stack(stk);
-    caml_raise(exn);
+    caml_raise(result.data);
```
I tried to think about style that feel less like it violates the abstraction of the `caml_result` type. `caml_run_result` (with a better name) would be better, I think; or an `exn = result.data` following immediately the test for `is_exception`, imitating the style used previously. This is only important here insofar as we want to think about which style one would recommend to programmers for resource clean-up.

gasche replied:
I went for the "binding immediately after the test" as a form of pattern-matching:

```c
  if (result.is_exception) {
    value exn = result.data;
    compare_free_stack(stk);
    caml_raise(exn);
  }
```

gadmm commented on runtime/fail_byt.c:
```diff
@@ -39,9 +39,10 @@ CAMLexport void caml_raise(value v)
+  {
+    caml_result result = caml_process_pending_actions_with_root_result(v);
+    v = result.data;
+  }
```
The blanket `v = result.data;` is very unusual so it could benefit from a comment saying that this is really intended.

It looks like one can do without the braces here.

gasche replied:
New version.

```c
  caml_result result = caml_process_pending_actions_with_root_result(v);
  /* If the result is a value, we want to assign it to [v].
     If the result is an exception, we want to continue raising it instead of [v].
     The line below does both these things at once. */
  v = result.data;
```

gadmm commented on runtime/fail_nat.c:
```diff
@@ -65,9 +65,10 @@ void caml_raise(value v)
+  {
+    caml_result result = caml_process_pending_actions_with_root_result(v);
+    v = result.data;
+  }
```
(idem)

gadmm reviewed (commented):
Overall I think it's a step in a good direction. I have some comments below.

Also, I think one important part remaining is to update the documentation (`intf-c.etex`), which should show that it is simpler indeed for living in harmony with the GC.

xavierleroy commented on runtime/caml/mlvalues.h:
```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
-/* Encoded exceptional return values, when functions are suffixed with
-   _exn. Encoded exceptions are invalid values and must not be seen
-   by the garbage collector. */
+/* A 'result' type/monad for OCaml computations. */
```
The "M" word is known to inspire terror in many programmers, cf. SPJ's "warm fuzzy things" alternate naming.  Maybe just "A 'result' type" ?

xavierleroy reviewed (commented):
Minor remarks below.  Overall, this looks good, but as discussed privately with @gasche I'd be reassured if this design was tested on the OCaml/Rust interface, to make sure it works well for this particularly demanding library.

gadmm commented:
This needs to be rebased on top of the statmemprof patches, and `memprof.c` adapted to the new `_res` convention.

gasche committed:
document and refactor [caml_process_pending_actions_result]

Suggested-by: Damien Doligez <damien.doligez@inria.fr>

gasche committed:
mlvalues.h: move encoded exceptions to the end to discourage usage

Suggested-by: Guillaume Munch-Maccagnoni <Guillaume.Munch-Maccagnoni@inria.fr>

gasche committed:
runtime: make caml_raise_if_exception private to callback.c

gasche committed:
runtime: rename caml_run_result in caml_get_value_or_raise

gasche committed:
runtime: remove encoded exceptions from memprof.c

gasche commented:
I rebased this PR on top of trunk, and took all the standing review comments into account. Thanks a lot for all this detailed feedback.

Notable changes:
- @damiendoligez suggested to simplify `caml_process_pending_actions_result`, this is done in a commit called "document and refactor [caml_process_pending_actions_result" (I am not giving specify hashes as they will be invalidated on my next rebase, so I recommend to look for it [in the list of commits](https://github.com/ocaml/ocaml/pull/13013/commits).
- as @gadmm pointed out, the newly merged memprof.c codebase used `_exn` functions quite a bit. I converted them to _result functions in "runtime: remove encoded exceptions from memprof.c", and this can be taken as an example of _result-heavy code. (@gadmm had some justified worries about the heaviness of the current naming conventions, maybe we can check that they are acceptable there)
- @gadmm wanted another name for `caml_run_result`, closer to `caml_raise_if_exception`. I moved `caml_raise_if_exception` (which was CAML_INTERNALS) out of the way, and renamed `caml_run_result` into `caml_get_value_or_raise` -- but my mind is at peace with the idea that I may need to change it again and again. Instead of a function exposed in mlvalues.h and defined in fail_{byt,nat}.c, this is now an inline function of fail.h.

Besides the naming of `caml_get_value_or_raise`, there are some questions left in the air about whether we want the conversion functions between `caml_result` and encoded exceptions to be made public or not.

gasche commented:
Of course I forgot two points.

I said that I would move from `_Bool` to `int`, but I am feeling nostalgic about a proper boolean type already -- and then maybe I need to use `intnat`, I am not sure. I am hoping that more knowledgeable people will conclusively say that one of these options is much better than the other.

@xavierleroy asked whether this `caml_result` design works in Rust. The answer is "yes, according to `bindgen`" (the standard Rust tool that takes C headers and generates compatible Rust definitions for them), as reported by @gadmm (quoting from his log):

Input: h.h
```c
typedef unsigned long long value;

typedef struct caml_result {
    _Bool is_exception;
    value data;
} caml_result;

caml_result caml_process_pending_actions_res();
```

Output: `bindgen h.h` (with some boring tests elided)

```rust
/* automatically generated by rust-bindgen 0.69.4 */

pub type value = ::std::os::raw::c_ulonglong;
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct caml_result {
    pub is_exception: bool,
    pub data: value,
}
#[test]
fn bindgen_test_layout_caml_result() {
    const UNINIT: ::std::mem::MaybeUninit<caml_result> = ::std::mem::MaybeUninit::uninit();
    let ptr = UNINIT.as_ptr();
    assert_eq!(
        ::std::mem::size_of::<caml_result>(),
        16usize,
        concat!("Size of: ", stringify!(caml_result))
    );
    assert_eq!(
        ::std::mem::align_of::<caml_result>(),
        8usize,
        concat!("Alignment of ", stringify!(caml_result))
    );
}
extern "C" {
    pub fn caml_process_pending_actions_res() -> caml_result;
}
```

This suggests that the `caml_result` type *does* have a representation compatible with Rust structures (when constrained by the `#[repr(C)]` pragma), so interoperability should work fine. This is not an *idiomatic* way to deal with results-or-failures in Rust (as a user I would expect a `Result<OCamlValue,OCamlValue>` type instead), but conversion functions to can be used on the Rust side if desired. We cannot hope to do better today as the standard `Result` type in Rust is not `repr(C)`, so there is no way to bind it directly from C in a compatible way. (While studying this I found the tool [cbindgen](https://github.com/mozilla/cbindgen/blob/master/docs.md) that supports repr(C) enums as [really tagged unions](https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md), but it does not apply here.)

xavierleroy commented:
> I said that I would move from _Bool to int, but I am feeling nostalgic about a proper boolean type already -- and then maybe I need to use intnat

`int` should work, as we don't have to guarantee size and alignment for the `caml_result` type.  I also thought of
```
typedef struct caml_result {
    enum { OK = 0, ERR = 1 } status;
    value data;
} caml_result;
```
which gets you even closer to the result type in Rust or OCaml.  Plus, it leaves open the possibility of adding other kinds of results later, such as `FATAL_ERROR = 2`, i.e. a result that `caml_get_value_or_raise` would turn into a fatal error.  Just an idea.

NickBarnes commented on runtime/callback.c:
```diff
@@ -294,27 +293,60 @@ CAMLexport value caml_callbackN_exn(value closure, int narg, value args[]) {
+CAMLexport caml_result caml_callback_result(
+  value closure, value arg)
+{
+  return Result_encoded(caml_callback_exn(closure, arg));
```
I suppose it would be more effort to make `caml_callback_exn` (etc) be a wrapper around `caml_callback_result`, rather than vice versa. But in future, since the "encoded" versions are frowned upon, I can imagine switching them over (and even, for that matter, pushing the change down into `caml_callback_asm`).

gasche replied:
The plan is to keep using the "encoded exceptions" format for code that is defined in the assembly runtime, because there it is much easier to flip a boolean than to return a structure. Now assuming that all the `caml_callback*_asm` functions return an encoded exception, it is more logical I thnink to have the `_exn` version be primitive and the `_result` version defined on top of it -- otherwise the `_exn` version would go back and forth, from encoded exceptions to caml_result to encoded exception again.

NickBarnes commented on runtime/callback.c:
```diff
@@ -294,27 +293,60 @@ CAMLexport value caml_callbackN_exn(value closure, int narg, value args[]) {
 CAMLexport value caml_callback (value closure, value arg)
 {
-  return caml_raise_if_exception(caml_callback_exn(closure, arg));
+  return encoded_value_or_raise(caml_callback_exn(closure, arg));
```
or `caml_get_value_or_raise(caml_callback_result(closure, arg))` ?

gadmm commented on runtime/caml/fail.h:
```diff
@@ -116,6 +114,19 @@ CAMLnoret CAMLextern void caml_array_bound_error (void);
 CAMLnoret CAMLextern void caml_raise_sys_blocked_io (void);
 
+
+#ifdef CAML_INTERNALS
```
I'd be in favour of making caml_get_value_or_raise non-internal—in fact the previous version of my related PR was making caml_raise_if_exception non-internal. Also I think fail.h is a good place for this.

gasche replied:
I did this change.

gadmm commented on runtime/caml/fail.h:
```diff
@@ -116,6 +114,19 @@ CAMLnoret CAMLextern void caml_array_bound_error (void);
+
+#ifdef CAML_INTERNALS
+/* Returns the value of a [caml_result] or raises the exception.
+   This function replaced [caml_raise_if_exception] in 5.3. */
```
No need to mention caml_raise_if_exception since it was internal

gasche replied:
I prefer to document than not, it could be useful for external users that do use CAML_INTERNALS and have to port their code to a new release, and I see no harm for everyone else.

gadmm commented on runtime/caml/memory.h:
```diff
@@ -405,6 +405,10 @@ struct caml__roots_block {
   }
 
+#define CAMLlocalresult(res) \
+  res.data = Val_unit; \
```
shouldn't this be
```suggestion
  caml_result res = Result_unit; \
```
or am I missing something?

gasche replied:
The `CAMLlocalresult` definition was wrong, and it was not used in the runtime. I believe that it is useful to have such a macro (it is tricky to write, and it is easy to write subtly wrong code by forgetting about it), so what I did was to (1) fix it and (2) tweak the code to use it somewhere (where it is not strictly necessary). As you have seen, #13061 is an offshoot of this work, and I may wait for that one to be resolved to decide what to do here.

I originally introduced `CAMLlocalresult` because I had a use for it in the runtime, but the previous use-case apparently was just in pre-PR experiments and never made it to the PR. It is notable that it is not strictly necessary in the runtime itself, where the cleanup code in the exception case does not require rooting, and the usage code in the value case is either (1)  an immediate return or (2) a continuation assuming that the value is `()` and never using it again.

NickBarnes reviewed (approved):
Looks good to me.

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -84,12 +84,34 @@ typedef opcode_t * code_t;
+   cleanup and propagate the exception themselves.
+
+   To extract the value or raise the exception, use
+     value caml_get_value_or_raise(caml_result)
```
You are indeed hinting at a non-internal definition of `caml_get_value_or_raise` (which I am in favour of)

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -471,6 +493,36 @@ CAMLextern value caml_set_oo_id(value obj);
         Caml_out_of_heap_header_with_reserved(wosize, tag, 0)
 
+
+/* Deprecated/discouraged suppport for unsafe encoded exceptions.
```
```suggestion
/* Obsolete -- support for unsafe "encoded" exceptions.
```
Deprecated is ambiguous as it has a specific meaning of "candidate for removal in a future version of the language" in some standardized languages. Since encoded exceptions were a public documented API and correct programs using it can be expected to exist, I think you mean obsolete/discouraged.

gadmm commented on runtime/caml/mlvalues.h:
```diff
@@ -471,6 +493,36 @@ CAMLextern value caml_set_oo_id(value obj);
+#define Is_exception_result(v) (((v) & 3) == 2)
+#define Extract_exception(v) ((v) & ~3)
+
+Caml_inline value Encoded_result(caml_result res) {
```
IIUC these would only be useful for programmers who are changing their programs to take advantage of the new feature. I agree that it's better to let them redefine these conversions themselves (which is straightforward).

gadmm commented on runtime/memprof.c:
```diff
@@ -2156,8 +2158,10 @@ void caml_memprof_sample_young(uintnat wosize, int from_caml,
 
-  if (cancelled)
-    caml_raise(Extract_exception(res));
+  if (res.is_exception) {
```
is the removal of `cancelled` intended?

gadmm commented on runtime/signals.c:
```diff
@@ -350,59 +349,64 @@ value caml_do_pending_actions_exn(void)
+CAMLexport caml_result caml_process_pending_actions_result(void)
 {
-  return caml_process_pending_actions_with_root_exn(Val_unit);
+  if (caml_check_pending_actions()) {
```
I don't see what's wrong with `return caml_process_pending_actions_with_root_result(Val_unit);`. It's better to not duplicate the code in this case.

gadmm commented:
@gasche has already explored the idea of having the result type extensible in the future. However I don't see how it fits with the stability guarantees needed.

I don't think there is a decisive argument for/against any of the suggested representations, just pick any one.

gasche committed:
review: remove Encoded_result, move Result_encoded to callback.c

gasche committed:
review: make is_exception 'int' to avoid _Bool compatibility issues

gasche committed:
gc_full_major: make it clearer that no rooting is necessary

gasche committed:
review: make caml_get_value_or_raise non-internal

gasche committed:
review: fix CAMLlocalresult and use it to ensure that it works

gasche commented:
I did a round of modification following the reviews, thanks!

For now I have used `int` for the `is_exception` field.

>  I also thought of
> 
> ```c
> typedef struct caml_result {
>     enum { OK = 0, ERR = 1 } status;
>     value data;
> } caml_result;
> ```
> 
> which gets you even closer to the result type in Rust or OCaml. Plus, it leaves open the possibility of adding other kinds of results later, such as `FATAL_ERROR = 2`, i.e. a result that `caml_get_value_or_raise` would turn into a fatal error. Just an idea.

I did, in a previous iteration of this PR, use an enum with two values: CAML_VALUE and CAML_EXCEPTION (I am not sure if enum names need namespacing in C, but I thought that they did). The result was fairly heavy/cumbersome, because I would then feel tempted to use `switch` on this field, and have to handle the default case, etc. Maybe I should revisit this with explicit equality tests instead of a switch.

The only form of extensibility that I could think of was an eventual future version of `caml_result` that could also represent unhandled effects. But this seemed like science-fiction, I doubted that the extension would be grateful, and I am skeptical that we would ever try to have C layers support continuation capture. I don't know about Fatal_error, but a closely related idea is Asynchronous_exception if the proposal to distinguish them from "normal exception" goes through.

What I would ideally like to do to have something more future-proof than the current definition is the following:

1. Expose a function to test if a result is an exception, instead of accessing the field directly. But I don't know how to name this function, in a way that would not conflict with `Is_exception_result`, which was an excellent name but is now taken for encoded exceptions. (I thought of reappropriating the name by removing the previous function, but that sounds a bit wild for something outside CAML_INTERNALS.)

2. Hide the definition of `caml_result` as a structure, to force users to go through the `Result_value` and `Result_exception` constructors, and I suppose some accessors as well. But I don't know how to do this in C in a way that does not hurt performance by breaking the inlining of the trivial wrappers.

gasche commented:
Also: @gadmm proposed changing the convention from `foo_result` to `foo_res`. His arguments make sense, so I am thinking of giving the rename a try to see how it goes in the code, but I haven't had the time to do so yet.

gadmm commented:
Thanks for the feedback, I was going to insist about it :) In the related PR, the _result suffix often leads to ambiguities
- caml_threadstatus_new_exn -> caml_threadstatus_new_result
- caml_thread_domain_initialize_hook_exn -> caml_thread_domain_initialize_hook_result
- caml_check_error_exn -> caml_check_error_result

gasche committed:
mlvalues.h: expose caml_result as an more abstract type

gasche committed:
caml_result: avoid using the 'private' fields directly when possible

gasche commented:
This PR has suffered from a decision paralysis on my side: I wasn't sure what definition of `caml_result` was the right definition to expose, and paralyzed by the fact that it would be publicly accessible and set in stone forever.

I just pushed the following change:
- the definition of mlvalues.h explicitly declares that the structure definition is private -- the type should be considered as opaque
- to let users use results without accessing the structure fields, a new `int caml_result_is_exception(caml_result)` function is offered
- this abstraction layer is now used almost everywhere in the runtime

Relevant code:
```c
typedef struct caml_result_private caml_result;

/* This structure should be considered internal, its definition may
   change in the future. Its public interface is formed of
   - Result_value, Result_exception
   - caml_result_is_exception
   - caml_get_value_or_raise (in fail.h)
*/
struct caml_result_private {
  int is_exception;
  value data;
};

#define Result_value(v) (struct caml_result_private){ .is_exception = 0, .data = v }
#define Result_exception(exn) (struct caml_result_private){ .is_exception = 1, .data = exn }

Caml_inline int caml_result_is_exception(struct caml_result_private result)
{
  return result.is_exception;
}
```

I am now at peace with the idea that an `int is_exception` field might not be the ideal definition to set in stone forever, and willing to move forward again with the PR.

My plan for now is to try the `_res` suffixing rename suggested by @gadmm, possibly push this rename if I like the result better, and then move toward merging the PR based on @damiendoligez's approval. Do not hesitate to speak up if you would like to see things done differently.

gasche committed:
runtime: rename all [foo_result] function into [foo_res]

Suggested-by: Guillaume Munch-Maccagnoni <Guillaume.Munch-Maccagnoni@inria.fr>

gasche commented:
I pushed a rename of the `foo_result` functions into `foo_res`. I don't have a strong opinion on which is better.

This has the downside of being harder to understand, which is arguably also an upside as it clearly marks the functions as following a specific convention. This avoid clashes with some functions which already used `_result` for other purposes, for example `sync_result` or `caml_lazy_read_result`.

Example of code following the new convention:

```c
  /* Call signal handlers first */
  caml_result result = caml_process_pending_signals_res();
  if (caml_result_is_exception(result)) goto exception;

  /* Call memprof callbacks */
  result = caml_memprof_run_callbacks_res();
  if (caml_result_is_exception(result)) goto exception;

  /* Call finalisers */
  result = caml_final_do_calls_res();
  if (caml_result_is_exception(result)) goto exception;
```

gadmm commented:
Note that saying that the definition is private does not necessarily mean that you can change it later. Adding a new case can still break programs. So I do not really see the point.

It is necessary to update the documentation, how do you see it being approached?

Thanks for changing to `_res`, this definitely reduces the risks of ambiguity.

gasche commented:
> It is necessary to update the documentation, how do you see it being approached?

Yes, I agree that I should probably update the documentation. I may try to prepare a draft PR to submit separately, but hold off from any merge-related action here before that documentation is at least submitted.

gasche commented:
I propose documentation for `caml_result` in the FFI chapter of the manual in #13216.

gasche commented:
Note to self: the Coq implementation uses `_exn` functions in its fork of the bytecode interpreter:

```
# /usr/bin/ld: kernel/byterun/libcoqrun_stubs.a(coq_interp.o): in function `coq_interprete':
# /home/gasche/Prog/ocaml/github-trunk/_opam/.opam-switch/build/coq-core.8.19.2/_build/default/kernel/byterun/coq_interp.c:582: undefined reference to `caml_process_pending_actions_exn'
```

To fix this, Coq would need conditional compilations depending on the OCaml version, or we would have to restore `caml_process_pending_actions_exn` for backwards-compatibility.

gadmm commented:
This is part of the public interface indeed. In fact, I checked that you preserved the functions part of the public interface so it did not raise alarm: `caml_process_pending_actions_exn` is still declared in `signals.h` indeed. It seems right to implement `caml_process_pending_actions_exn` in terms of `caml_process_pending_actions_res` for backwards-compatibility.

gasche commented:
I restore `caml_process_pending_actions_exn` in #13304.
